### PR TITLE
Peer2Peer Networking

### PR DESCRIPTION
**Pull request by coot**

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 1]
```diff
@@ -43,6 +43,9 @@ import           Data.Void (Void)
 import           Data.ByteString.Lazy (ByteString)
 import           Data.Kind (Type)
 import           System.Random (newStdGen, split)
+#ifdef POSIX
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:61]: Add a todo that this can be removed once `trace-forwarder` is merged

File: ouroboros-network/src/Ouroboros/Network/Diffusion.hs [hunk 2]
```diff
@@ -59,25 +74,31 @@ data ExtraArguments (p2p :: P2P) where
 
 -- | Application data which depend on p2p mode.
 --
-data ExtraApplications (p2p :: P2P) where
+data ExtraApplications (p2p :: P2P) n2nAddr where
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion.hs]: Rename to `ntnAddr`

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Simple.hs [hunk 3]
```diff
@@ -74,28 +83,26 @@ withPeerSelectionActions
...
     -- (for example because the node hasn't synced far enough) we fall back
     -- to using the manually configured bootstrap root peers.
-    requestPublicRootPeers :: DNSActions DNS.Resolver IOException IO
-                           -> Int -> IO (Set peeraddr, DiffTime)
-    requestPublicRootPeers dnsActions n = do
-      peers_m <- getLedgerPeers (NumberOfPeers $ fromIntegral n)
-      case peers_m of
-           Nothing    -> requestConfiguredRootPeers dnsActions n
-           Just peers -> return peers
+    requestLedgerPeers :: Int -> m (Set peeraddr, DiffTime)
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Simple.hs:94]: Preserve the original name.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Simple.hs [hunk 4]
```diff
@@ -74,28 +83,26 @@ withPeerSelectionActions
...
+             Nothing    -> requestPublicRootPeers n
+             Just peers -> return peers
 
     -- For each call we re-initialise the dns library which forces reading
     -- `/etc/resolv.conf`:
     -- https://github.com/input-output-hk/cardano-node/issues/731
-    requestConfiguredRootPeers :: DNSActions DNS.Resolver IOException IO
-                               -> Int -> IO (Set peeraddr, DiffTime)
-    requestConfiguredRootPeers dnsActions n =
+    requestPublicRootPeers :: Int -> m (Set peeraddr, DiffTime)
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Simple.hs:104]: Preserve the original name.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 5]
```diff
@@ -379,47 +403,105 @@ runWith RunNodeArgs{..} LowLevelRunNodeArgs{..} =
...
+               ByteString
+               ByteString
+               ByteString
+               ()
+        )
       -> (   BlockNodeToClientVersion blk
           -> NodeToClientVersion
-          -> NTC.Apps m (ConnectionId addrNTC) ByteString ByteString ByteString ()
+          -> NTC.Apps
+               m (ConnectionId ntcAddr) ByteString ByteString ByteString ()
```
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:422]: We should use `addrNTC` here.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 6]
```diff
@@ -379,47 +403,105 @@ runWith RunNodeArgs{..} LowLevelRunNodeArgs{..} =
...
     mkDiffusionApplications
-      :: MiniProtocolParameters
+      :: NetworkP2PMode p2p
+      -> MiniProtocolParameters
       -> (   BlockNodeToNodeVersion blk
-          -> NTN.Apps m (ConnectionId addrNTN) ByteString ByteString ByteString ByteString ByteString ()
-         )
+          -> NTN.Apps
+               m
+               (ConnectionId ntnAddr)
```
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:411]: We should use `addrNTN`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 7]
```diff
@@ -379,47 +403,105 @@ runWith RunNodeArgs{..} LowLevelRunNodeArgs{..} =
...
+      peerMetrics
+      btime =
+      case enP2P of
+        EnabledP2PMode ->
+          ( Diffusion.Applications {
+              Diffusion.daApplicationInitiatorMode = initiator,
+              Diffusion.daApplicationInitiatorResponderMode = initiatorAndResponder,
+              Diffusion.daLocalResponderApplication = responder,
+              Diffusion.daLedgerPeersCtx = ledgerPeersConsensusInterface
+            }
```
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:448]: Define `Diffusion.Applications` for both p2p and non-p2p in `where` clause.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 8]
```diff
@@ -709,7 +805,7 @@ stdLowLevelRunNodeArgsIO RunNodeArgs{ rnProtocolInfo } StdRunNodeArgs{..} = do
       , llrnVersionDataNTN =
           stdVersionDataNTN
             networkMagic
-            (daDiffusionMode srnDiffusionArguments)
+            (Diffusion.daDiffusionMode srnDiffusionArguments)
```
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:808]: Maybe we could just use `daMode`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 9]
```diff
@@ -379,47 +403,105 @@ runWith RunNodeArgs{..} LowLevelRunNodeArgs{..} =
...
+              Diffusion.daLocalResponderApplication = responder,
+              Diffusion.daLedgerPeersCtx = ledgerPeersConsensusInterface
+            }
+          , Diffusion.P2PApplications
+              P2P.ApplicationsExtra {
+                P2P.daMiniProtocolParameters = miniProtocolParams,
+                P2P.daRethrowPolicy = consensusRethrowPolicy (Proxy @blk),
+                P2P.daLocalRethrowPolicy = localRethrowPolicy,
+                P2P.daPeerMetrics = peerMetrics,
+                P2P.daBlockFetchMode = getFetchMode (getChainDB kernel) btime
```
Comment (coot) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:455]: Is it possible to get `getFetchMode` from `BlockFetchConsensusInterface`?

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs [hunk 10]
```diff
@@ -364,12 +364,7 @@ pattern CardanoNodeToClientVersion7 =
 instance CardanoHardForkConstraints c
       => SupportedNetworkProtocolVersion (CardanoBlock c) where
   supportedNodeToNodeVersions _ = Map.fromList $
-      [ (NodeToNodeV_1, CardanoNodeToNodeVersion1)
-      , (NodeToNodeV_2, CardanoNodeToNodeVersion2)
-      , (NodeToNodeV_3, CardanoNodeToNodeVersion2)
-      , (NodeToNodeV_4, CardanoNodeToNodeVersion3)
-      , (NodeToNodeV_5, CardanoNodeToNodeVersion4)
-      , (NodeToNodeV_6, CardanoNodeToNodeVersion4)
+      [ (NodeToNodeV_6, CardanoNodeToNodeVersion4)
```
Comment (coot) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:367]: Let's talk with 3rd party tools that it will break them.

File: network-mux/src/Network/Mux.hs [hunk 11]
```diff
@@ -214,6 +215,7 @@ runMux :: forall m mode.
        -> m ()
 runMux tracer Mux {muxMiniProtocols, muxControlCmdQueue, muxStatus} bearer = do
     egressQueue <- atomically $ newTBQueue 100
+    labelTBQueueIO egressQueue "mux-eq"
```
Comment (karknu) [inline network-mux/src/Network/Mux.hs:218]: TODO: this should be named <bearer> ++ "mux-eq"

File: ouroboros-network-framework/src/Ouroboros/Network/Snocket.hs [hunk 12]
```diff
@@ -104,11 +104,12 @@ import           Ouroboros.Network.Linger (StructLinger (..))
 -- descriptor by `createNamedPipe`, see 'namedPipeSnocket'.
 --
 newtype Accept m fd addr = Accept
-  { runAccept :: m (Accepted fd addr, Accept m fd addr)
+  { runAccept :: (forall a. m a -> m a) -- ^ unmask exceptions
+              -> m (Accepted fd addr, Accept m fd addr)
```
Comment (karknu) [inline ouroboros-network-framework/src/Ouroboros/Network/Snocket.hs:108]: Don't change the Accept type.

File: ouroboros-network-framework/src/Simulation/Network/Snocket.hs [hunk 13]
```diff
@@ -23,6 +23,7 @@ module Simulation.Network.Snocket
     withSnocket
   , ResourceException (..)
   , SnocketTrace (..)
+  , TimeoutDetail (..)
```
Comment (karknu) [inline ouroboros-network-framework/src/Simulation/Network/Snocket.hs:26]: This seems to belong to another commit. `snocket: better mask/unmask snocket and socket resources`.

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs [hunk 14]
```diff
@@ -11,7 +11,6 @@ module Network.Mux.Bearer.AttenuatedChannel
   , Attenuation (..)
   , newConnectedAttenuatedChannelPair
   , attenuationChannelAsMuxBearer
-  , CloseError (..)
```
Comment (karknu) [inline network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs:14]: TODO: should be moved out of mux.

File: ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Unversioned.hs [hunk 15]
```diff
@@ -8,6 +8,9 @@ module Ouroboros.Network.Protocol.Handshake.Unversioned
   , unversionedHandshakeCodec
   , unversionedProtocolDataCodec
   , unversionedProtocol
+  , DataFlowProtocolData (..)
```
Comment (karknu) [inline ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Unversioned.hs:11]: TODO: this file belongs in a testing framework so that it isn't accidentally used by real code.

File: ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs [hunk 16]
```diff
@@ -1563,8 +1563,10 @@ withConnectionManager ConnectionManagerArguments {
                 return (DemoteToColdLocalNoop Nothing
                                               (abstractState $ Known connState))
 
-              InboundIdleState _connId _connThread _handle dataFlow ->
-                assert (dataFlow == Duplex) $
+              -- TODO: This assertion is benign and also hit rarely (once per
+              -- 100_000 simulations)
```
Comment (karknu) [inline ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs:1826]: Hi marcin!

File: ouroboros-network-framework/src/Ouroboros/Network/Server2.hs [hunk 17]
```diff
@@ -176,9 +174,24 @@ run ServerArguments {
...
     iseCONNABORTED :: IOError -> Bool
+#if defined(mingw32_HOST_OS)
+    -- On Windows the network packet classifies all errors
+    -- as OtherError. This means that we're forced to match
+    -- on the error string. The text string comes from
+    -- the network package's winSockErr.c, and if it ever
+    -- changes we must update our text string too.
+    iseCONNABORTED (IOError _ _ _ "Software caused connection abort (WSAECONNABORTED)" _ _) = True
+    iseCONNABORTED _ = False
+#else
```
Comment (karknu) [inline ouroboros-network-framework/src/Ouroboros/Network/Server2.hs:186]: TODO: create ticket where we move eCONNABORTED detection to Snocket accept and push the osx workaround to network package.

File: network-mux/src/Network/Mux.hs (hunk 11 above)
Comment (coot) [reply network-mux/src/Network/Mux.hs:218]: Issue [3472](https://github.com/input-output-hk/ouroboros-network/issues/3472).

File: network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs (hunk 14 above)
Comment (coot) [reply network-mux/src/Network/Mux/Bearer/AttenuatedChannel.hs:14]: https://github.com/input-output-hk/ouroboros-network/issues/3473

File: ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Unversioned.hs (hunk 15 above)
Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Unversioned.hs:11]: https://github.com/input-output-hk/ouroboros-network/issues/3474

File: ouroboros-network-framework/src/Ouroboros/Network/Server2.hs (hunk 17 above)
Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/Server2.hs:186]: https://github.com/input-output-hk/ouroboros-network/issues/3475

File: ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs (hunk 16 above)
Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs:1826]: If the assertion fires, this means we are deregistering an outbound connection which is in `InboundIdleState`.  The `Server2` tests do not have a proper outbound governor, which can lead to this kind of problem.  It is benign, because if we try to unregistered a connection which already was unregistered, we do not need to do anything.  Still, it'd be nice to fix this, so we can have a stronger confidence that outbound governor / peer state actions, do not try to do these kind of things.

Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs:1826]: https://github.com/input-output-hk/ouroboros-network/issues/3477

**Review by karknu — APPROVED** [review]
LGTM.
Great job Marcin! Turning p2p-master into something that is mergeable was a monumental task.

**Review by dcoutts — APPROVED** [review]
:beers:

**Comment by coot** [comment]
Thanks @karknu, @dcoutts, @bolt12 and @avieth for all the contributions.

**Comment by coot** [comment]
bors merge

**Review by nfrisby — COMMENTED** [review]
LGTM thanks for the ping.
